File: java-practice/debezium-embedded/src/main/java/com/demo/application/DebeziumListener.java
package com.demo.application;

import com.demo.repository.TbTargetRepository;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.debezium.engine.ChangeEvent;
import io.debezium.engine.DebeziumEngine;
import io.debezium.engine.format.Json;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.Properties;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

@Service
public class DebeziumListener {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private final Properties config;
    private final Executor executor = Executors.newSingleThreadExecutor();
    private final TbTargetRepository tbTargetRepository;

    public DebeziumListener(
        @Qualifier("connector") Properties config,
        TbTargetRepository tbTargetRepository
    ) {
        this.config = config;
        this.tbTargetRepository = tbTargetRepository;
    }

//    @PostConstruct
    public void start() {
        executor.execute(() -> {
            try (DebeziumEngine<ChangeEvent<String, String>> engine = DebeziumEngine.create(Json.class)
                .using(config)
                .notifying(event -> {
                    System.out.println("📣 CDC Event:");
                    System.out.println(event.value());
                    var e = event.headers();
                    var ee = event.destination();
                    var key = event.key();

                    var target = event.value();

                    try {
                        Map<String, Object> map = MAPPER.readValue(target, new TypeReference<>() {
                        });

                        System.out.println(map);
                        System.out.println(map);
                        System.out.println(map);
                    } catch (JsonProcessingException ex) {
                        System.err.println("--------------------!");
                    }

                })
                .build()) {
                engine.run();
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }
}

File: api-practice/debezium-embedded/src/main/kotlin/lkdcode/server/debezium/listener/DebeziumListener.kt
package lkdcode.server.debezium.listener

import io.debezium.engine.ChangeEvent
import io.debezium.engine.DebeziumEngine
import io.debezium.engine.format.Json
import jakarta.annotation.PostConstruct
import jakarta.annotation.PreDestroy
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import java.util.*
import java.util.concurrent.Executors


@Service
class DebeziumListener(
    @Qualifier("debeziumConnector")
    private val properties: Properties
) {
    private val executor = Executors.newSingleThreadExecutor { r ->
        Thread(r, "debezium-lkdcode").apply { isDaemon = false }
    }

    private lateinit var engine: DebeziumEngine<ChangeEvent<String, String>>

    @PostConstruct
    fun start() {
        engine = DebeziumEngine.create(Json::class.java)
            .using(properties)
            .notifying { records: List<ChangeEvent<String, String>>,
                         committer: DebeziumEngine.RecordCommitter<ChangeEvent<String, String>> ->
                records.forEach { record ->
                    try {
                        println("CDC destination: ${record.destination()}")
                        println("CDC key: ${record.key()}")
                        println("CDC value: ${record.value()}")

                        committer.markProcessed(record)
                    } catch (e: Exception) {
                        println("ERROR: $e")
                    }
                }

                committer.markBatchFinished()
            }
            .build()

        executor.submit(engine)
    }

    @PreDestroy
    fun stop() {
        try {
            if (this::engine.isInitialized) engine.close()
        } finally {
            executor.shutdown()
        }
    }
}